services/scheduler/src/Scheduler.js

```diff
@@ -0,0 +1,136 @@
+        await this._queueCreator.assertMessagesQueue(messagesQueue, exchangeName, deadLetterRoutingKey);
+        await this._channel.sendToQueue(
+            this._queueCreator.getAmqpStepConfig(flow, flow.getFirstNode().id).messagesQueue,
+            Buffer.from(JSON.stringify(msg)),
+            {
+                headers:  scheduleRecord
+            }
+        );
+    }
+    async _loopBody() {
+        const flows = (await this._crdClient.flows.get()).body.items;
+        const schedulerRecords = (await this._crdClient.schedulerrecords.get()).body.items;
+        const schedulerRecordsIndex = schedulerRecords.reduce((index, record) => {
+            index[record.metadata.name] = record;
+            return index; 
+        }, {});
+        const now = new Date();
+        for (let flow of flows) {
+            const flowModel = new Flow(flow);
+            const firstStep = flowModel.getFirstNode();
+            if (flowModel.metadata.deletionTimestamp) {
+                this._logger.trace({flowId: flowModel.id}, 'flow is deleting now, skip');
+                continue; 
+            }
+            if (!firstStep || !firstStep.isPolling) {
+                this._logger.trace({flowId: flowModel.id}, 'flow is not polling, skip');
+                const schedulerRecord = schedulerRecordsIndex[flowModel.id];
+                if (schedulerRecord) { 
+                    this._logger.trace({
+                        name: schedulerRecord.metadata.name,
+                        dueExecution: schedulerRecord.spec.dueExecution
+                    }, 'Going to delete scheduler record due non-polling flow');
+                    await this._crdClient.schedulerrecords(flowModel.id).delete();
+                }
+                continue;
+            }
+            const schedulerRecord = schedulerRecordsIndex[flowModel.id] || {
+                "apiVersion": "elastic.io/v1",
+                "kind": "SchedulerRecord",
+                "metadata": {
+                    "name": flowModel.id,
+                    "namespace": this._config.get('NAMESPACE'),
+                    ownerReferences: [
+                        {
+                            apiVersion: "elastic.io/v1",                                     
+                            kind: "Flow",
+                            controller: true,
+                            name: flowModel.metadata.name,
+                            uid: flowModel.metadata.uid
+                        }
+                    ]
+                },
+                "spec": {
+                }
+            };
+            delete schedulerRecordsIndex[flowModel.id];
+            if (!schedulerRecord.spec.dueExecution || now >= new Date(schedulerRecord.spec.dueExecution)) { 
+                try {
+                    await this._scheduleOne(flowModel);
+
+                    const newDueExecution = new Date();
+                    newDueExecution.setMinutes(newDueExecution.getMinutes() + 3);
```

Comment:
Would be nice to move magic number to some constant.

Reply:
And how about `cron` flows?

Reply:
cron flows are not implemented atm